src/addrman.cpp

```diff
@@ -713,13 +713,18 @@ CAddrInfo CAddrMan::Select_(bool newOnly) const
         // use a tried node
         double fChanceFactor = 1.0;
         while (1) {
+            // Pick a tried bucket, and an initial position in that bucket.
             int nKBucket = insecure_rand.randrange(ADDRMAN_TRIED_BUCKET_COUNT);
             int nKBucketPos = insecure_rand.randrange(ADDRMAN_BUCKET_SIZE);
-            while (vvTried[nKBucket][nKBucketPos] == -1) {
-                nKBucket = (nKBucket + insecure_rand.randbits(ADDRMAN_TRIED_BUCKET_COUNT_LOG2)) % ADDRMAN_TRIED_BUCKET_COUNT;
-                nKBucketPos = (nKBucketPos + insecure_rand.randbits(ADDRMAN_BUCKET_SIZE_LOG2)) % ADDRMAN_BUCKET_SIZE;
+            // Iterate over the positions of that bucket.
+            int i;
+            for (i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+                if (vvTried[nKBucket][nKBucketPos ^ i] != -1) break;
```

Comment:
It seems to me this operation will iterate over all the elements of a bucket, but not serially. So if a bucket is non empty it will always find an address in that bucket.

Is the purpose of the above Xor is to add randomness into the search?

> That does mean that if a bucket has fewer entries, its entries are supposed to be picked more frequently.

I am also not understanding how the search will ensure the above? It seems irrespective of the bucket size, it will always find an entry if the bucket is non-empty.

Reply:
The xor is just a simple way of iterating over the positions in the bucket. It's not more or less predictable than starting in a random position and incrementing until it wraps around and reaches the starting position again - just less code for doing something similar.

> I am also not understanding how the search will ensure the above? It seems irrespective of the bucket size, it will always find an entry if the bucket is non-empty.

That's exactly what it does. The old code will pick a uniformly random bucket/position combination, among all of those that are non-empty. The new code will pick a uniformly random non-empty *bucket*, and then a random position in that bucket. That means that relatively speaking, if bucket A has 1 entry, and bucket B as 4 entries, and all other buckets are empty, A's entry will have 50% chance of being picked, and B's entries will have 12.5% each. In the old code, each would have 20%.

Reply:
Thanks @sipa for clarifying. Its natural that if probability of a bucket being chosen is `P(bucket)` and it has `n` addresses, then probability of a particular address being picked from a particular bucket are `P(bucket) / n`. That satisfies your example above.

But to me, that doesn't seem like satisfying your above statement, " if a bucket has fewer entries, its entries are supposed to be picked more frequently."

Because bucket choices are uniformly random, for `m` non-empty buckets, probability of a particular bucket being chosen is `1/m`. And, then because the search ensures that a non-empty bucket will always yield an address, so probability of getting **any** address from that bucket is `1`. 

Following this logic, we have, for `m` buckets, probability that an address from a particular bucket to be chosen is always `1/m`, uniformly distributed over all the buckets, and independent of size of buckets.

Another way to put it would be, just because the bucket has lesser address, it doesn't have any higher probability of being chosen, than rest of the buckets.

Does that make sense?

Reply:
> So mapping back to your example, it would say, probability of choosing an address from Bucket A == Probability of choosing an address from Bucket B == 50%. The address count doesn't play a role anymore, because we will always choose an address from a bucket if it has any.

Right, that's the case. But that is exactly the same as saying that addresses in buckets with more addresses are relatively less frequently going to be picked.

Reply:
But I do understand your point, that probability of a particular address being chosen from a particular bucket, decreases with number of addresses in that bucket. 
But that's just natural consequence, and the code doesn't favor smaller buckets in any way. Which I thought was the case by your statement.

Reply:
Ah, no my statement, reformulated, is: "if you look at the resulting change of per-address probabilities caused by this PR, you'll see it now favors those in smaller buckets more. However, that is desirable."